### PR TITLE
[runtime-security] do not cache file with hardlinks 

### PR DESCRIPTION
**Pull request by safchain**
### What does this PR do?

Do not cache path for file with hardlink so that the path resolution works with multiple path to a same file, meaning same inode. With this fix, busybox executables are now correctly resolved.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

Write here in detail how you have tested your changes
and instructions on how this should be tested in QA.

Describe or link instructions to set up environment 
for testing, if the process requires more than just
running the agent on one of the supported platforms.

### Checklist

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

File: pkg/security/ebpf/c/link.h [hunk 1]
```diff
@@ -105,14 +106,22 @@ int __attribute__((always_inline)) sys_link_ret(void *ctx, int retval, int dr_ty
...
-    syscall->resolver.iteration = 0;
-    syscall->resolver.ret = 0;
+    // invalidate user face inode, so no need to bump the discarder revision in the event
+    if (retval >= 0) {
+        // for hardlink we need to invalidate the cache as the nlink counter in now > 1
+        invalidate_inode(ctx, syscall->link.src_file.path_key.mount_id, syscall->link.src_file.path_key.ino, !is_event_enabled(EVENT_LINK));
+    }
 
-    resolve_dentry(ctx, dr_type);
+    if (!syscall->discarded && is_event_enabled(EVENT_LINK)) {
```
Comment (lebauce) [inline pkg/security/ebpf/c/link.h:115]:
```
int pass_to_userspace = !syscall->discarded && is_event_enabled(EVENT_LINK)
invalidate_inode(ctx, syscall->link.src_file.path_key.mount_id, syscall->link.src_file.path_key.ino, !pass_to_userspace);
```

File: pkg/security/ebpf/c/rename.h [hunk 2]
```diff
@@ -116,7 +116,7 @@ int __attribute__((always_inline)) sys_rename_ret(void *ctx, int retval, int dr_
 
     // invalidate user face inode, so no need to bump the discarder revision in the event
     if (retval >= 0) {
-        invalidate_inode(ctx, syscall->rename.target_file.path_key.mount_id, syscall->rename.target_file.path_key.ino, 1);
+        invalidate_inode(ctx, syscall->rename.target_file.path_key.mount_id, syscall->rename.target_file.path_key.ino, !is_event_enabled(EVENT_RENAME));
```
Comment (lebauce) [inline pkg/security/ebpf/c/rename.h:119]:
```
int pass_to_userspace = !syscall->discarded && is_event_enabled(EVENT_RENAME)
invalidate_inode(ctx, syscall->link.src_file.path_key.mount_id, syscall->link.src_file.path_key.ino, !pass_to_userspace);
```

File: pkg/security/probe/probe.go [hunk 3]
```diff
@@ -441,6 +441,12 @@ func (p *Probe) handleEvent(CPU uint64, data []byte) {
 			log.Errorf("failed to decode link event: %s (offset %d, len %d)", err, offset, dataLen)
 			return
 		}
+
+		// need to invalidate as now nlink > 1
+		if event.Link.Retval >= 0 {
+			// defer it do ensure that it will be done after the dispatch that could re-add it
+			defer p.invalidateDentry(event.Link.Source.MountID, event.Link.Source.Inode)
```
Comment (lebauce) [inline pkg/security/probe/probe.go:448]: Isn't the invalidation of the dentry done twice without the suggested changed related to pass_to_userspace ?

File: pkg/security/tests/hardlink_test.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	"fmt"
+	"os"
+	"os/exec"
+	"testing"
+
+	sprobe "github.com/DataDog/datadog-agent/pkg/security/probe"
+	"github.com/DataDog/datadog-agent/pkg/security/rules"
+)
+
+func TestHardLink(t *testing.T) {
```
Comment (Gui774ume) [inline pkg/security/tests/hardlink_test.go:20]: We should test both the eRPC method and the Map method separately.

**Review by Gui774ume — COMMENTED** [review]
Left a few comments, but overall it looks great !
(FYI I think you committed an empty file 🤔  )

**Review by Gui774ume — APPROVED** [review]
lgtm

**Review by Gui774ume — APPROVED** [review]

**Pull request merged**